internal/cover/cover.go

```diff
@@ -0,0 +1,218 @@
+			panic("non-int index in coverData: " + k)
+		}
+		vi, err := strconv.Atoi(v)
+		if err != nil {
+			panic("non-int value in coverData: " + v)
+		}
+		res[ki] = vi
+	}
+	return res
+}
+
+func (cov *coverageHelper) annotateActions(actions []*ast.Action) (res []*ast.Action) {
+	for _, action := range actions {
+		action.Stmts = cov.annotateStmts(action.Stmts)
+		res = append(res, action)
+	}
+	return
+}
+
+func (cov *coverageHelper) annotateFunctions(functions []*ast.Function) (res []*ast.Function) {
+	for _, function := range functions {
+		function.Body = cov.annotateStmts(function.Body)
+		res = append(res, function)
+	}
+	return
+}
+
+func (cov *coverageHelper) annotateStmtsList(stmtsList []ast.Stmts) (res []ast.Stmts) {
+	for _, stmts := range stmtsList {
+		res = append(res, cov.annotateStmts(stmts))
+	}
+	return
+}
+
+// annotateStmts takes a list of statements and adds counters to the beginning of
+// each basic block at the top level of that list. For instance, given
+//
+//	S1
+//	if cond {
+//		S2
+//	}
+//	S3
+//
+// counters will be added before S1,S2,S3.
+func (cov *coverageHelper) annotateStmts(stmts ast.Stmts) (res ast.Stmts) {
+	var simpleStatements []ast.Stmt
+	for _, stmt := range stmts {
+		wasBlock := true
+		switch s := stmt.(type) {
+		case *ast.IfStmt:
+			s.Body = cov.annotateStmts(s.Body)
+			s.Else = cov.annotateStmts(s.Else)
+		case *ast.ForStmt:
+			s.Body = cov.annotateStmts(s.Body) // TODO should we do smth with pre & post?
+		case *ast.ForInStmt:
+			s.Body = cov.annotateStmts(s.Body)
+		case *ast.WhileStmt:
+			s.Body = cov.annotateStmts(s.Body)
+		case *ast.DoWhileStmt:
+			s.Body = cov.annotateStmts(s.Body)
+		case *ast.BlockStmt:
+			s.Body = cov.annotateStmts(s.Body)
+		default:
+			wasBlock = false
+		}
+		if wasBlock {
+			if len(simpleStatements) > 0 {
+				res = append(res, cov.trackStatement(simpleStatements))
+				res = append(res, simpleStatements...)
+			}
+			res = append(res, stmt)
+			simpleStatements = []ast.Stmt{}
+		} else {
+			simpleStatements = append(simpleStatements, stmt)
+		}
+	}
+	if len(simpleStatements) > 0 {
+		res = append(res, cov.trackStatement(simpleStatements))
+		res = append(res, simpleStatements...)
+	}
+	return
+	// TODO complete handling of if/else/else if
+}
+func (cov *coverageHelper) trackStatement(statements []ast.Stmt) ast.Stmt {
+	op := "=1"
+	if cov.covermode == "count" {
+		op = "++"
+	}
+	cov.annotationIdx++
+	start1, _ := statements[0].(ast.BoundaryProvider).Boundary()
+	_, end2 := statements[len(statements)-1].(ast.BoundaryProvider).Boundary()
+	path, startLine := cov.fileReader.FileLine(start1.Line)
+	_, endLine := cov.fileReader.FileLine(end2.Line)
+	cov.boundaries[cov.annotationIdx] = boundary{
+		start:    Position{startLine, start1.Column},
+		end:      Position{endLine, end2.Column},
+		fileName: path,
+	}
+	cov.stmtsCnt[cov.annotationIdx] = len(statements)
+	return parseProg(fmt.Sprintf(`BEGIN { %s[%d]%s }`, ArrCover, cov.annotationIdx, op)).Begin[0][0]
```

Comment:
No offense, but my variant is cleaner :). It's 1 line with simple code, easy to understand what AST we want to generate from a string of code. Yours is 6+ lines and covers only one case of two. 
Besides, this is pretty standard and established approach in such cases. I learned this trick implementing AWK plugin for IntellJ IDEA. Please take a look at code example from JetBrains people: https://github.com/JetBrains/intellij-community/blob/0e2aa4030ee763c9b0c828f0b5119f4cdcc66f35/java/java-psi-impl/src/com/intellij/psi/impl/PsiElementFactoryImpl.java#L428.

Reply:
Your version is definitely more succinct, however, I found it confusing at first, as in "wow, why is he firing up the parser here? and what's this magical-looking `.Begin[0][0]` doing?" Took me a minute to figure out, "ah, I see, he's using that to build a parse tree". Efficiency is not a great concern here, but it's also quite a heavy operation, *per statement*: lex, parse, resolve, compile.

I'd feel better about this approach if the compiling was done for two statements once. But that approach would lead to some messy fiddling with the AST to get the dynamic values in.

I know mine is significantly longer (though we will get rid of the `parseProg` helper), but I think it's very straight-forward and obvious -- it looks like building an AST, and it *is* building an AST -- as well as efficient.

You could factor out the `Left`, as that's the same for assignment and incrementing, something like so:

```go
left := &ast.IndexExpr{
	Array: ast.ArrayRef(ArrCover, Position{}),
	Index: []ast.Expr{&ast.NumExpr{Value: float64(cov.annotationIdx)}},
}
if cov.covermode == "count" {
	return &ast.ExprStmt{Expr: &ast.IncrExpr{Expr: left, Op: INCR}}
}
return &ast.ExprStmt{Expr: &ast.AssignExpr{Left: left, Right: &ast.NumExpr{Value: 1}}}
```

Reply:
Regarding performance. How many nanoseconds will it help to save with your variant vs mine? Is it relevant for all goawk invocations?

Regarding clarity. Definitelly, your variant is clearer in terms of what it's doing (building AST). But to me it is less clear in terms of what exact AST is it building. Where in my case once you've got the idea, it's obvious that we are building the AST for AWK code `%s[%d]%s` filled with `ArrCover, cov.annotationIdx, op`. If clarity is not enough - we can just add a comment above this line.

That being said, by no means am I opposed to your proposal. If after reading the two items above you still disagree - I'll apply your change with no further arguing.

Reply:
Yeah, I still disagree about this. Let's please go with the AST but add a comment, something like this:

```go
left := &ast.IndexExpr{
	Array: ast.ArrayRef(ArrCover, Position{}),
	Index: []ast.Expr{&ast.NumExpr{Value: float64(cov.annotationIdx)}},
}
if cov.covermode == "count" {
	// AST for __COVER[index]++
	return &ast.ExprStmt{Expr: &ast.IncrExpr{Expr: left, Op: INCR}}
}
// AST for __COVER[index] = 1
return &ast.ExprStmt{Expr: &ast.AssignExpr{Left: left, Right: &ast.NumExpr{Value: 1}}}
```

Reply:
Done.